### PR TITLE
Feat/ascii: Integer lib updated

### PR DESCRIPTION
**Pull request by 0xKahi**
## Pull Request type
The ASCII Integer Library serves as a practical tool for converting Cairo's integer types (`u8`, `u16`, `u32`, `u64`, `u128`, `u256`) into their ASCII representations as felt252.

as requested from #101 
- the code has become more generic removing repeats

``` cairo
#[test]
#[available_gas(2000000000)]
fn u256_to_ascii() {
    // ------------------------------ max u256 test ----------------------------- //
    // max u256 int in cairo is GitHub Copilot: The maximum u256 number in Cairo is `
    // 115792089237316195423570985008687907853269984665640564039457584007913129639935`.
    let num: u256 = BoundedInt::max();
    let ascii: Array<felt252> = num.to_ascii();
    assert(ascii.len() == 3, 'max u256 wrong len');
    assert(*ascii.at(0) == '1157920892373161954235709850086', 'max u256 wrong first felt');
    assert(*ascii.at(1) == '8790785326998466564056403945758', 'max u256 wrong second felt');
    assert(*ascii.at(2) == '4007913129639935', 'max u256 wrong third felt');
    // ------------------------------ min u256 test ----------------------------- //
    let num: u256 = BoundedInt::min();
    let ascii: Array<felt252> = num.to_ascii();

    assert(ascii.len() == 1, 'min u256 wrong len');
    assert(*ascii.at(0) == '0', 'min u256 wrong felt');
    // ---------------------------- 31 char u256 test --------------------------- //
    let ascii: Array<felt252> = 1157920892373161954235709850086_u256.to_ascii();
    assert(ascii.len() == 1, 'u256 31 char wrong len');
    assert(*ascii.at(0) == '1157920892373161954235709850086', '31 char u256 wrong felt');
    // ---------------------------- 62 cahr u256 test --------------------------- //
    let ascii: Array<felt252> = 11579208923731619542357098500868790785326998466564056403945758_u256.to_ascii();
    assert(ascii.len() == 2, 'u256 31 char wrong len');
    assert(*ascii.at(0) == '1157920892373161954235709850086', '31 char u256 wrong felt');
    assert(*ascii.at(1) == '8790785326998466564056403945758', '62 char u256 wrong felt');
}
```

Please check the type of change your PR introduces:

- [x] Feature

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/ascii/interger.cairo [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            num = quotient;
+        };
+        new_arr
+    }
+}
+
+impl U256ToAsciiTraitImpl of IntergerToAsciiTrait<u256, Array<felt252>> {
+
+    fn to_ascii(self: u256) -> Array<felt252> {
+        let mut data = ArrayTrait::<felt252>::new();
```
Comment (gaetbout) [inline src/ascii/interger.cairo:185]: No need to specify ::<felt252> is it inferred

File: src/ascii/interger.cairo [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+> of IntergerToAsciiArrayTrait<T> {
+    fn to_ascii_array(self: T) -> Array<felt252> {
+        let mut new_arr = self.to_inverse_ascii_array();
+        new_arr.reverse();
+        new_arr
+    }
+
+    fn to_inverse_ascii_array(self: T) -> Array<felt252> {
+        let mut new_arr = ArrayTrait::<felt252>::new();
+        if self <= TryInto::<felt252, T>::try_into(9).unwrap() {
```
Comment (gaetbout) [inline src/ascii/interger.cairo:39]: can do:
```suggestion
        if self <= 9.try_into().unwrap() {
```

File: src/ascii/interger.cairo [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            new_arr.append(self.into() + 48);
+            return new_arr;
+        }
+
+        let mut num = self;
+        loop {
+            if num <= Zeroable::<T>::zero() {
```
Comment (gaetbout) [inline src/ascii/interger.cairo:46]: Do you mean num.is_zero() ?

Comment (0xKahi) [reply src/ascii/interger.cairo]: yep i have pushed the new changes thanks

File: src/ascii/interger.cairo [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+            return new_arr;
+        }
+
+        let mut num = self;
+        loop {
+            if num.is_zero() {
+                break ();
+            }
+            let (quotient, remainder) = DivRem::div_rem(
+                num, TryInto::<felt252, T>::try_into(10).unwrap().try_into().expect('Division by 0')
```
Comment (gaetbout) [inline src/ascii/interger.cairo:50]: missing a try_into here :)

File: src/ascii/interger.cairo [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        let mut index = 0;
+        let mut ascii: felt252 = 0;
+        loop {
+            if index >= len {
+                // if ascii is 0 it means we have already appended the first ascii
+                // and theres no need to append it again
+                match ascii {
+                    0 => (),
+                    _ => data.append(ascii),
+                }
```
Comment (gaetbout) [inline src/ascii/interger.cairo:202]: This could be a if ascii.is_non_zero()

File: src/ascii/interger.cairo (hunk 4 above)
Comment (0xKahi) [reply src/ascii/interger.cairo:50]: i cant seem to use try_into() like that without specifying the type as it fails to infer the trait

Comment (gaetbout) [reply src/ascii/interger.cairo:50]: you prob can force the type by doing `10.try_into::<felt, T>().unwrap()`

Comment (0xKahi) [reply src/ascii/interger.cairo:50]: hmm dosent seem to work the error is `Wrong number of generic arguments. Expected 0, found: 2`

File: src/ascii/interger.cairo (hunk 5 above)
Comment (0xKahi) [reply src/ascii/interger.cairo]: yep mad the changes thanks

File: src/ascii/interger.cairo [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        let mut ascii: felt252 = 0;
+        loop {
+            if index >= len {
+                break ();
+            }
+            // recursively keep getting the index from the end of the array
+            let l_index = len - index - 1;
+            ascii = ascii * 256 + *inverse_ascii_arr[l_index];
+            index += 1;
+        };
```
Comment (gaetbout) [inline src/ascii/interger.cairo:91]: You'll need to update imports `use array::{ArrayTrait, SpanTrait};`

This part could be updated to:
```suggestion
        let mut inverse_ascii_arr = (self.to_inverse_ascii_array()).span();
        let mut ascii: felt252 = 0;
        loop {
          match inverse_ascii_arr.pop_back() {
              Option::Some(val) => {
                  ascii = ascii * 256 + *val;
              },
              Option::None(_) => {
                  break;
              },
          };
```

File: src/ascii/interger.cairo [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+        if self <= 9.try_into().unwrap() {
+            data.append(self.into() + 48);
+            return data;
+        }
+
+        let inverse_ascii_arr = self.to_inverse_ascii_array();
+        let len = inverse_ascii_arr.len();
+        let mut index = 0;
+        let mut ascii: felt252 = 0;
+        loop {
```
Comment (gaetbout) [inline src/ascii/interger.cairo:122]: I think the pop_back pattern can also be applied here

File: src/ascii/interger.cairo [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        if self <= 9 {
+            data.append(self.try_into().expect('number overflow felt252') + 48);
+            return data;
+        }
+
+        let inverse_ascii_arr = self.to_inverse_ascii_array();
+        let len = inverse_ascii_arr.len();
+        let mut index = 0;
+        let mut ascii: felt252 = 0;
+        loop {
```
Comment (gaetbout) [inline src/ascii/interger.cairo:194]: And alos here

File: src/ascii/interger.cairo [hunk 9]
```diff
@@ -0,0 +1,220 @@
+use core::option::OptionTrait;
+use alexandria::data_structures::array_ext::ArrayTraitExt;
+use array::{ArrayTrait, Array};
+use traits::{Into, TryInto, DivRem};
+use zeroable::Zeroable;
+
+trait IntergerToAsciiTrait<T, U> {
```
Comment (LucasLvy) [inline src/ascii/interger.cairo:7]:
```suggestion
trait ToAsciiTrait<T, U> {
```

File: src/ascii/interger.cairo [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    fn to_ascii(self: T) -> U;
+}
+
+// converts intergers into an array of its individual ascii values
+trait IntergerToAsciiArrayTrait<T> {
```
Comment (LucasLvy) [inline src/ascii/interger.cairo:12]:
```suggestion
trait ToAsciiArrayTrait<T> {
```

File: src/ascii/interger.cairo [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    fn to_ascii_array(self: T) -> Array<felt252>;
+    fn to_inverse_ascii_array(self: T) -> Array<felt252>;
+}
+
+// converts intergers into an array of its individual ascii values
+// e.g. 123 -> [49, 50, 51]
+impl IntergerToAsciiArrayTraitImpl<
+    T,
+    impl TNumericLiteral: NumericLiteral<T>,
```
Comment (LucasLvy) [inline src/ascii/interger.cairo:21]: is that needed ?

File: src/ascii/interger.cairo [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+> of IntergerToAsciiArrayTrait<T> {
+    fn to_ascii_array(self: T) -> Array<felt252> {
+        let mut new_arr = self.to_inverse_ascii_array();
+        new_arr.reverse();
+        new_arr
+    }
+
+    fn to_inverse_ascii_array(self: T) -> Array<felt252> {
+        let mut new_arr = ArrayTrait::new();
+        if self <= 9.try_into().unwrap() {
```
Comment (LucasLvy) [inline src/ascii/interger.cairo:39]: can't you just use into() ?

File: src/ascii/interger.cairo [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+            new_arr.append(self.into() + 48);
+            return new_arr;
+        }
+
+        let mut num = self;
+        loop {
+            if num.is_zero() {
+                break ();
```
Comment (LucasLvy) [inline src/ascii/interger.cairo:47]:
```suggestion
                break;
```

File: src/ascii/interger.cairo [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+        new_arr
+    }
+}
+
+// gneric implementation for small intergers <u128 
+// to transform its intergers into a string represented as a single felt252
+// e.g. 1000 -> "1000"
+impl SmallIntergerToAsciiTraitImpl<
+    T,
+    impl TNumericLiteral: NumericLiteral<T>,
```
Comment (LucasLvy) [inline src/ascii/interger.cairo:64]: is that needed ?

File: src/ascii/interger.cairo [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+            return self.into() + 48;
+        }
+
+        let inverse_ascii_arr = self.to_inverse_ascii_array();
+        let len = inverse_ascii_arr.len();
+        let mut index = 0;
+        let mut ascii: felt252 = 0;
+        loop {
+            if index >= len {
+                break ();
```
Comment (LucasLvy) [inline src/ascii/interger.cairo:85]:
```suggestion
                break;
```

File: src/ascii/interger.cairo [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+        ascii
+    }
+}
+
+// gneric implementation for big intergers u128 
+// to transform its intergers into a string represented as multiple felt252 if there is overflow
+// e.g. max_num + 123 -> ["max_num", "123"]
+impl BigIntergerToAsciiTraitImpl<
+    T,
+    impl TNumericLiteral: NumericLiteral<T>,
```
Comment (LucasLvy) [inline src/ascii/interger.cairo:101]: is that needed ?

File: src/ascii/interger.cairo [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+        let mut ascii: felt252 = 0;
+        loop {
+            if index >= len {
+                // if ascii is 0 it means we have already appended the first ascii
+                // and theres no need to append it again
+                if ascii.is_non_zero() {
+                    data.append(ascii);
+                }
+
+                break ();
```
Comment (LucasLvy) [inline src/ascii/interger.cairo:130]:
```suggestion
                break;
```

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: src/ascii/interger.cairo (hunk 16 above)
Comment (LucasLvy) [reply src/ascii/interger.cairo:101]: i was talking about this line `impl TNumericLiteral: NumericLiteral<T>,`

File: src/ascii/interger.cairo (hunk 12 above)
Comment (0xKahi) [reply src/ascii/interger.cairo:39]: unfortunately no if i were to bring in a generic impl of `Into<felt252, T>` the trait would not work on uints as they do not have a felt252 -> uint Into trait implementation on them only a try_into trait

Comment (LucasLvy) [reply src/ascii/interger.cairo:39]: you have
`impl TInto: Into<T, felt252>`
just do 
`if self.into() <= 9_felt252` ?

File: src/ascii/interger.cairo (hunk 16 above)
Comment (0xKahi) [reply src/ascii/interger.cairo:101]: oh wait i just realise you are talking about the TNumericLiteral generic impl il remove it and test it out

File: src/ascii/interger.cairo (hunk 12 above)
Comment (0xKahi) [reply src/ascii/interger.cairo:39]: the `PartialOrd` impl is set to type `T` when i tried that i get this error `Trait has no implementation in context: core::traits::PartialOrd::<core::felt252>`

Comment (LucasLvy) [reply src/ascii/interger.cairo:38]: ah yes then keep it as is

**Review by LucasLvy — APPROVED** [review]
Maybe we could change the traits to use `ref self` instead of self to avoid moving it but it's not blocking

**Comment by LucasLvy** [comment]
@0xKahi are you on telegram?

**Comment by 0xKahi** [comment]
> @0xKahi are you on telegram?

yep @J0xkujo is my tele

**Pull request merged**